Apply rollup.js for bundling

banyan opened the pull request:
Tried to solve using rollup for https://github.com/florian/cookie.js/issues/41. 

Prepared 3 patterns: 

* https://github.com/banyan/issue-repro/tree/cookie_js-41-a
  * this is using `import * as cookie from 'cookie_js';`
  * `yarn start` and `yarn test` don't work.
* https://github.com/banyan/issue-repro/tree/cookie_js-41-b
  * this is using `import { cookie } from 'cookie_js';`
  * while `yarn test` works, `yarn start` doesn't work
* https://github.com/banyan/issue-repro/tree/cookie_js-41-c
  * this branch is using this PR. `yarn start` and `yarn test` both works.

banyan commented:
@florian 

Can you please review? I'll also accept if you want me to change anything. 🙇

florian commented on tests/index.html:
```diff
@@ -20,8 +20,8 @@
-	<script src="../cookie.js"></script>
-	<script src="spec.js"></script>
+	<script src="../dist/cookie.umd.js"></script>
+	<script src="./spec.js"></script>
```
Does the `./` make any difference? It looks like it should work without.

banyan replied:
Thanks! Updated: https://github.com/florian/cookie.js/pull/47/commits/1447fc5f1b16e2fb50866850b40e0175de3c6f22

florian replied:
Thanks!

florian commented on src/cookie.js:
```diff
@@ -0,0 +1,169 @@
+// Copyright (c) 2015 Florian Hartmann, https://github.com/florian https://github.com/florian/cookie.js
```
Can you use `git mv cookie.js src/cookie.js` instead of just `mv`?
That'd make the diffs easier to review and we wouldn't lose any git history on that file
Assuming you haven't made any changes to that file itself, it should also fix the merge conflict.

banyan replied:
Thanks! Updated: https://github.com/florian/cookie.js/pull/47/commits/5ff762ac9d5cde2cc9023442c34e22fdd8c4b811

florian replied:
Thanks!

florian commented on tests/builds/cjs.js:
```diff
@@ -0,0 +1,9 @@
+var assert = require('assert');
```
Importing spec.js, and running it again here and in the esm test would be perfect.
It might be a bit much for this PR though.
If it's too much work, I'll try to do it myself later. :)

banyan replied:
I think this requires some tasks a bit. so I'd like to possible to avoid in this PR 🙇

florian replied:
Sure thing, sounds good to me :)

banyan commented:
hmm, test is somehow flaky. (works on my end though) let me investigate a bit..

florian commented:
Thanks for the changes!

It also works locally for me if that's of any help.
The travis errors do look pretty weird though, I wouldn't expect `cookie.utils.isArray` to stop working after all these years, especially since the function is fairly self-contained.

banyan commented:
at least, I was able to reproduce it locally using the following command.

```
for i in {0..100};do npm test; echo $? >> `date "+%Y-%m-%d"`.txt; done
```

* 1st attempt against node v10: failed 22 out of 100.
* 2nd attempt against node v10: failed 30 out of 100.
* 1st attempt against node v6: failed 36 out of 100.
* 2nd attempt against node v6: failed 29 out of 100.

What I can see here is that the version of the node is irrelevant.

Most of the failed patterns are like this.

```
  ․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․
․!․
```

```
  ․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․
․․․․!!
```

But sometimes it's broken quite randomly.

```
  ․․․․․․․․․․․․․․․․․․․․․․․․․․․․․
․!․․!․․!․!․!․․!․!․!․!․․!․!․
```

so I think fundamental layer is somehow wrong.
the library uses phantomjs. I added a [commit](https://github.com/florian/cookie.js/pull/47/commits/48829537699b75dc1a6ebe697f8a75dbf12b1155) otherwise phantomjs is not detected in my local.
I'm not sure if that's caused or not, but I ended up not to use phantomjs since it's not maintained.

I used jsdom instead of phantomjs and it became pretty stable.
Also, I added some changes.

* Add test for each module building as you suggested
* Add minified build for compatibility if someone uses
* Remove grunt since grunt is mainly used for grunt-mocha and it's not used now
  * I've removed jshint but I can add eslint if you prefer later this PR

I haven't changed the core code basically, but I'm sorry that the PR has become bigger :bow:

florian commented:
Thanks so much! This is an incredible PR. 🎉 🙏
Also thanks a lot for improving the test cases further, this is awesome!
And again sorry for letting you wait for so long.

I've merged the PR and will push the updated version to package managers soon.

---

On another note, it's a bit frustrating how short-lived some of these JS tools are.
When I originally created the repo, it was in a really good state.
I had to rewrite a lot of the grunt / mocha stuff a few years in, but it seems like a lot of it was totally outdated now. :/
It's especially frustrating if install rules / commands just stop working together, e.g. the phantom stuff.

Thanks a lot for bringing everything up to date again! :)

banyan commented:
@florian 

Thanks for taking time for the review and also understanding the transition from phantomjs to jsdom 🙇 

>it's a bit frustrating how short-lived some of these JS tools are.
When I originally created the repo, it was in a really good state.
I had to rewrite a lot of the grunt / mocha stuff a few years in, but it seems like a lot of it was totally outdated now. :/
It's especially frustrating if install rules / commands just stop working together, e.g. the phantom stuff.

True! 😂
